File: 20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start

import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char ch : s.toCharArray()) {
            if (ch == '(') {
                stack.add(')');
            }
            if (ch == '[') {
                stack.add(']');
            }
            if (ch == '{') {
                stack.add('}');
            }
            if (ch == ')' || ch == ']' || ch == '}') {
                if (stack.size() == 0) {
                    return false;
                }
                if (stack.pop() != ch) {
                    return false;
                }
            }
        }
        return (stack.size() == 0);
    }
}
// @lc code=end

File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start

#include <stack>
#include <string>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (char ch : s) {
            if (ch == '(') { st.push(')'); continue; }
            if (ch == '[') { st.push(']'); continue; }
            if (ch == '{') { st.push('}'); continue; }
            if (st.empty()) { return false; }
            if (st.top() != ch) { return false; }
            st.pop();
        }
        return st.empty();
    }
};
// @lc code=end
